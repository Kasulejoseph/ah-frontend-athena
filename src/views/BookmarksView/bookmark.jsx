import React, { Component } from "react";
import PropTypes from "prop-types";
import { MDBIcon } from "mdbreact";
import { connect } from "react-redux";
import { BookmarkAction } from "../../actions/bookmarkAction/BookmarkAction";
import { UnBookmarkAction } from "../../actions/bookmarkAction/UnbookmarkAction";
import { getMyBookmarkedArticlesAction } from "../../actions/getMyBookmarkedArticlesAction";
import ModalPage from "../../components/Likes/LoginModal";

import "./bookmarks.scss";

const token = window.localStorage.getItem("token");
export class Bookmark extends Component {
  state = {
    is_bookmarked: false,
    bookmark: {},
    colorValue: "blue-grey-text ml-2",
    modal: false
  };

  componentDidMount() {
    const { getMyBookmarkedArticlesAction } = this.props;
    getMyBookmarkedArticlesAction(token);
  }

  componentWillReceiveProps(nextProps) {
    const { data, bookmarkChanges } = nextProps;
    const { getMyBookmarkedArticlesAction } = this.props;
    const { articleId } = this.props;
    if (data.bookmark) {
      this.setState({
        colorValue: "blue-grey-text ml-2",
        is_bookmarked: false
      });
      data.bookmark.map(bookmark => {
        if (bookmark.article === articleId) {
          return this.setState({
            bookmark: bookmark,
            colorValue: "blue-text pr-3",
            is_bookmarked: true
          });
        } else {
          return undefined;
        }
      });
    }
    if (bookmarkChanges) {
      getMyBookmarkedArticlesAction(token);
    }
  }

  toggle = () => {
    const { modal } = this.state;
    this.setState({
      modal: !modal
    });
  };

  onClick = () => {
    const { BookmarkAction, UnBookmarkAction, slug } = this.props;
    const { is_bookmarked, bookmark } = this.state;

    if (window.localStorage.getItem("token") != null) {
      if (is_bookmarked) {
        UnBookmarkAction(bookmark.id);
        getMyBookmarkedArticlesAction(token);
      } else {
        BookmarkAction(bookmark, slug);
        getMyBookmarkedArticlesAction(token);
      }
    } else {
      this.setState({ url: window.location.pathname });
      this.toggle();
    }
  };

  render() {
    const { colorValue, url, modal } = this.state;
    const { history } = this.props;

    return (
      <div className={colorValue}>
        <MDBIcon
          size="2x"
          onClick={this.onClick}
          icon="bookmark"
          className="bookmark-icon"
        />
        <ModalPage
          title="Please login before you can like or dislike the article"
          modal={modal}
          toggle={this.toggle}
          fallback={url}
          history={history}
          md="12"
        />
      </div>
    );
  }
}

Bookmark.propTypes = {
  getMyBookmarkedArticlesAction: PropTypes.func.isRequired,
  BookmarkAction: PropTypes.func.isRequired,
  bookmarkChanges: PropTypes.func.isRequired,
  UnBookmarkAction: PropTypes.func.isRequired,
  articleId: PropTypes.number,
  data: PropTypes.shape({}),
  slug: PropTypes.string,
  history: PropTypes.shape({})
};
Bookmark.defaultProps = {
  articleId: 2,
  data: {},
  slug: "mocked-SLUG",
  history: {}
};
export const mapStateToProps = state => {
  return {
    data: state.bookmarks.bookmarksData,
    bookmarkChanges: state.bookmarks.refresh
  };
};
export default connect(
  mapStateToProps,
  { BookmarkAction, UnBookmarkAction, getMyBookmarkedArticlesAction }
)(Bookmark);
