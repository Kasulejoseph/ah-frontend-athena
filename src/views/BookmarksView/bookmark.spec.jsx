import React from "react";
import { shallow } from "enzyme";
import { Bookmark, mapStateToProps } from "./bookmark";

let props = {
  message: "Article was bookmarked",
  is_bookmarked: false,
  BookmarkAction: jest.fn(),
  getMyBookmarkedArticlesAction: jest.fn(),
  UnBookmarkAction: jest.fn()
};
const initialState = {
  bookmarks: { bookmarksData: "mock bookmarks data" },
  bookmarksChanges: {}
};

describe("Testing the BookmarksView component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Bookmark {...props} />);
  });

  it("should compare the component with the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle props", () => {
    wrapper.setProps({ ...props });
    expect(wrapper.instance().props.message).toEqual("Article was bookmarked");
  });

  it("should handle onclick when isbookmark set to false", (id = jest.fn()) => {
    wrapper.setProps({ ...props });
    wrapper.instance().onClick(id);
    expect(wrapper.state("is_bookmarked")).toEqual(false);
  });

  it("should handle onclick when isbookmark set to true", (id = jest.fn()) => {
    props.is_bookmarked = true;
    wrapper.setProps({ ...props });
    wrapper.instance().onClick(id);
    expect(wrapper.state("is_bookmarked")).toEqual(false);
  });

  it("should handle the case when bookmarked articles data is available", () => {
    wrapper.setProps({
      bookmark: { bookmark: [{ article: "data" }] }
    });
    wrapper.setState({ is_bookmarked: false });
    expect(wrapper.state("is_bookmarked")).toEqual(false);
  });

  it("should map state to props", () => {
    expect(mapStateToProps(initialState).data).toEqual("mock bookmarks data");
  });
});
