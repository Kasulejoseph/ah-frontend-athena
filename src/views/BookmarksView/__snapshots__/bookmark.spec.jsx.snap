// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the BookmarksView component should compare the component with the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Bookmark
    BookmarkAction={[MockFunction]}
    UnBookmarkAction={[MockFunction]}
    articleId={2}
    data={Object {}}
    getMyBookmarkedArticlesAction={
      [MockFunction] {
        "calls": Array [
          Array [
            null,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    history={Object {}}
    is_bookmarked={false}
    message="Article was bookmarked"
    slug="mocked-SLUG"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <t
          border={false}
          brand={false}
          className="bookmark-icon"
          fab={false}
          fal={false}
          far={false}
          fixed={false}
          flip=""
          icon="bookmark"
          inverse=""
          light={false}
          list={false}
          onClick={[Function]}
          pull=""
          pulse={false}
          regular={false}
          rotate=""
          size="2x"
          spin={false}
          stack=""
        />,
        <ModalPage
          history={Object {}}
          md="12"
          modal={false}
          title="Please login before you can like or dislike the article"
          toggle={[Function]}
        />,
      ],
      "className": "blue-grey-text ml-2",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "border": false,
          "brand": false,
          "className": "bookmark-icon",
          "fab": false,
          "fal": false,
          "far": false,
          "fixed": false,
          "flip": "",
          "icon": "bookmark",
          "inverse": "",
          "light": false,
          "list": false,
          "onClick": [Function],
          "pull": "",
          "pulse": false,
          "regular": false,
          "rotate": "",
          "size": "2x",
          "spin": false,
          "stack": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fallback": undefined,
          "history": Object {},
          "md": "12",
          "modal": false,
          "title": "Please login before you can like or dislike the article",
          "toggle": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <t
            border={false}
            brand={false}
            className="bookmark-icon"
            fab={false}
            fal={false}
            far={false}
            fixed={false}
            flip=""
            icon="bookmark"
            inverse=""
            light={false}
            list={false}
            onClick={[Function]}
            pull=""
            pulse={false}
            regular={false}
            rotate=""
            size="2x"
            spin={false}
            stack=""
          />,
          <ModalPage
            history={Object {}}
            md="12"
            modal={false}
            title="Please login before you can like or dislike the article"
            toggle={[Function]}
          />,
        ],
        "className": "blue-grey-text ml-2",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "border": false,
            "brand": false,
            "className": "bookmark-icon",
            "fab": false,
            "fal": false,
            "far": false,
            "fixed": false,
            "flip": "",
            "icon": "bookmark",
            "inverse": "",
            "light": false,
            "list": false,
            "onClick": [Function],
            "pull": "",
            "pulse": false,
            "regular": false,
            "rotate": "",
            "size": "2x",
            "spin": false,
            "stack": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fallback": undefined,
            "history": Object {},
            "md": "12",
            "modal": false,
            "title": "Please login before you can like or dislike the article",
            "toggle": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
